aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -74,13 +79,16 @@
   private final AwsProperties awsProperties;
 
   private CountingOutputStream stream;
-  private final List<File> stagingFiles = Lists.newArrayList();
+  private final List<FileAndDigest> stagingFiles = Lists.newArrayList();
   private final File stagingDirectory;
   private File currentStagingFile;
   private String multipartUploadId;
   private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
   private final int multiPartSize;
   private final int multiPartThresholdSize;
+  private final boolean isEtagCheckEnabled;
+  private final MessageDigest completeMessageDigest;
+  private final MessageDigest currentPartMessageDigest;
```

Comment:
It looks like the purpose of having both digests is to handle the cases where either multi-part or PUT is used. Right?

Do we need both, or can we reuse the first part message digest?

Reply:
We need both to handle case where written bytes is larger than multipart size but less than multipart threshold. In that case we will have two files, and so two part message digests, whereas we need the whole message digest in put request.

Reply:
Can we recalculate the digest if we switch to multipart? Or can we at least stop using the complete message digest after switching to multipart?

Reply:
Yea, I was thinking about this too. We can't do that if we were to use digest stream, I will have to revert to my previous implementation of maintaining and updating digest in writes though.

Reply:
I don't think we want to try to recalculate if it requires rereading the file in order to calculate the checksum.  I still feel like pushing the per file digest into the FileAndDigest option is the best approach and keep the full file digest for the put object upload as well.  I think we're over optimizing by trying to stop the full file calculation and we could always log the full file MD5.  

There's another option that might simplify things but would require a larger change involving uploads.  If we simply make the first part of the multipart threshold larger (multiplart+threshold) then each FileAndDigest would have a digest associated with it and could be used for the put object or put part uploads.

Reply:
Actually, just realized we can just turn off the digest updates in `DigestOutputStream` after we have crossed multi part threshold size. Will add that.